src/appstatus.cpp

```diff
@@ -0,0 +1,171 @@
+#include "appstatus.h"
+
+AppStatus::AppStatus(QObject *parent)
+    : QObject(parent)
+    , m_status(Disabled)
+    , m_downloadStatus()
+    , m_uploadStatus()
+    , m_updating()
+    , m_bluetoothBusy()
+    , m_bluetoothEnabled()
+    , m_atRisk()
+    , m_downloadAvailable()
+{
+}
+
+AppStatus::Status AppStatus::status() const
+{
+    return m_status;
+}
+
+void AppStatus::updateStatus()
+{
+    Status status = m_status;
+
+    if (m_updating) {
+        m_status = Updating;
+    }
+    else if ((m_uploadStatus != Upload::StatusIdle) && (m_uploadStatus != Upload::StatusError)) {
+        m_status = Uploading;
+    }
+    else if ((m_downloadStatus != Download::StatusIdle) && (m_downloadStatus != Download::StatusError)) {
+        m_status = Downloading;
+    }
+    else if (m_uploadStatus == Upload::StatusError) {
+        m_status = ErrorUploading;
+    }
+    else if (m_downloadStatus == Download::StatusError) {
+        m_status = ErrorDownloading;
+    }
+    else if (m_bluetoothBusy) {
+        m_status = BluetoothBusy;
+    }
+    else if (m_atRisk) {
+        m_status = AtRisk;
+    }
+    else if (m_downloadAvailable) {
+        m_status = DailyUpdateRequired;
+    }
+    else if (m_bluetoothEnabled) {
+        m_status = Active;
+    }
+    else {
+        m_status = Disabled;
+    }
+
+    if (m_status != status) {
+        emit statusChanged();
+    }
+}
+
+Download::Status AppStatus::downloadStatus() const
+{
+    return m_downloadStatus;
+}
+
+Upload::Status AppStatus::uploadStatus() const
+{
+    return m_uploadStatus;
+}
+
+bool AppStatus::updating() const
+{
+    return m_updating;
+}
+
+bool AppStatus::bluetoothBusy() const
+{
+    return m_bluetoothBusy;
+}
+
+bool AppStatus::bluetoothEnabled() const
+{
+    return m_bluetoothEnabled;
+}
+
+bool AppStatus::atRisk() const
+{
+    return m_atRisk;
+}
+
+bool AppStatus::downloadAvailable() const
+{
+    return m_downloadAvailable;
+}
+
+void AppStatus::setDownloadStatus(Download::Status downloadStatus)
+{
+    if (m_downloadStatus != downloadStatus) {
+        m_downloadStatus = downloadStatus;
+        emit downloadStatusChanged();
+        updateStatus();
+        if (downloadStatus == Download::StatusError) {
+            emit notifyDownloadError();
+        }
+    }
+}
+
+void AppStatus::setUploadStatus(Upload::Status uploadStatus)
+{
+    if (m_uploadStatus != uploadStatus) {
+        m_uploadStatus = uploadStatus;
+        emit uploadStatusChanged();
+        updateStatus();
+        if (uploadStatus == Upload::StatusError) {
+            emit notifyUploadError();
+        }
+    }
+}
+
+void AppStatus::setUpdating(bool updating)
+{
+    if (m_updating != updating) {
+        m_updating = updating;
+        emit updatingChanged();
+        updateStatus();
```

Comment:
Same here with `updateStatus()`

Reply:
Or does it even matter? Are the AppStatus methods already called in seperate threads? Then, nevermind.

Reply:
Ah, are you thinking that this would avoid having the code run in the UI thread? In this case, yes, you're right, connecting these as signals/slots wouldn't actually affect things here.